### ISSUE TITLE
SQS visiblity period not using custom default value

### ISSUE THREAD

**Issue by fadilafidina** [Particular/NServiceBus.AmazonSQS#1158]
Hi,

We are using NServiceBus with SQS, without compatibility mode.

We think that all of our messages are defaulting to have a 30s visibility period timeout, even though we have set up a timeout of 2 mins 30 seconds from the AWS console itself for our queue.

We have a message handled in a saga that takes around 60-90 seconds to process, and we are seeing it get processed multiple times. We think it's because of the visibility period defaulting to 30 seconds, therefore the message was visible again on the queue, hence the multiple processing of the same message. However, we expect it to be using the 2 minutes 30 seconds setting that we have configured.

We see the same message being handled multiple times in quick succession, 0.1 of a second apart.

Our saga is set up for pessimistic locking.

Is there a reason why it's not adhering to the viability timeout we've configured?

We've also read that in https://docs.particular.net/nservicebus/sagas/concurrency#high-load-scenarios-avoid-io-or-cpu-bound-operations-in-a-saga-handler, we should not be processing long operations in a saga handler. Data contention is not a concern at the time that this process is executed.

**Comment by mauroservienti** [comment]
@fadilafidina thanks for raising this. I transferred the issue to the appropriate repository.

What version of the transport are you using?

You are correct, we are not honoring any message visibility timeout value defined on the main input queue. In the meantime, a workaround that might work for you is to enable the [message-driven pub/sub compatibility](https://docs.particular.net/transports/sqs/configuration-options#message-driven-pubsub-compatibility-mode) mode (via the `EnableMessageDrivenPubSubCompatibilityMode` config option) even if you don't need it, and set the message visibility timeout there:

```
var migrationSettings = transport.EnableMessageDrivenPubSubCompatibilityMode();
migrationSettings.MessageVisibilityTimeout(timeoutInSeconds: 300);
``` 

There might be a little performance penalty at publishing time if the endpoint publishes a lot of events. However, the built-in topics and subscriptions cache should compensate for that.

---

@danielmarbach this sounds like a nice-to-have improvement. IIRC we already retrieve information from the queues adding one more attribute should not be a big deal.

**Comment by fadilafidina** [comment]
Thanks for your prompt reply @mauroservienti! And thanks for moving the issue across - I realised afterwards that it should move across. 

We are using 5.4.0. 

We just removed compatibility mode from our code and have deployed it across most of our services, with the last bit of it being very soon. It would be great if we don't have to add it back in since it's being deprecated.

**Comment by danielmarbach** [comment]
@mauroservienti yes, I think that part of the code is still the same since the inception of the transport. I agree, adhering to the visibility timeout of the input queue would be a good improvement.

**Comment by mauroservienti** [comment]
Thanks, @danielmarbach. I investigated a bit, and I have a feeling it's a bug we introduced in 5.4. Here is my reasoning:

On `master` we do the following to start receiving  messages: https://github.com/Particular/NServiceBus.AmazonSQS/blob/0e8b4e30f2f80aa5ef9029bbda9404e901bbd7d7/src/NServiceBus.Transport.SQS/InputQueuePump.cs#L100-L107

We don't specify a `VisibilityTimeout` in the `ReceiveMessageRequest`. However, in `5.4` with the recent changes, the code now looks like this:  https://github.com/Particular/NServiceBus.AmazonSQS/blob/e3d310fc3477c6cf0d8910eef2b55a2daf7bb357/src/NServiceBus.Transport.SQS/InputQueuePump.cs#L81-L89

And due to the way the `MessageVisibilityTimeout` https://github.com/Particular/NServiceBus.AmazonSQS/blob/e3d310fc3477c6cf0d8910eef2b55a2daf7bb357/src/NServiceBus.Transport.SQS/TransportConfiguration.cs#L332-L343

We end up always specifying the visibility timeout for the `ReceiveMessageRequest`. My gut feeling is that `master`, by not setting a value, honors the value specified on the queue attributes, and `5.4` doesn't.

If my reasoning is correct, this qualifies as a bug.

**Comment by fadilafidina** [comment]
Hi @mauroservienti, @danielmarbach, thanks for looking into it and jumping on it quickly.

I see you're opened a draft PR on this - will not setting the value in `ReceiveMessageRequest` mean that it'll use the visibility timeout value that's set in SQS? That'd be our ideal solution.

**Comment by danielmarbach** [comment]
@mauroservienti yes it is a regression introduced in 5.4

> You can provide the VisibilityTimeout parameter in your request. The parameter is applied to the messages that Amazon SQS returns in the response. If you don't include the parameter, the overall visibility timeout for the queue is used for the returned messages. For more information, see [Visibility Timeout](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html) in the Amazon SQS Developer Guide. 

https://docs.aws.amazon.com/sdkfornet/v3/apidocs/items/SQS/TReceiveMessageRequest.html

This is how the transport was initially designed for. The setting was applied to the queue only, and then we honored whatever was on the queue, and therefore we did not specify the VisibilityTimeout on the receive request. Do you recall the reason why you introduced it on the request too?

**Comment by mauroservienti** [comment]
> Do you recall the reason why you introduced it on the request too?

It was meant to allow to specify a different visibility timeout for endpoints using the pub/sub compatibility mode in high throughput scenarios.

**Comment by mauroservienti** [comment]
@fadilafidina, we triaged it as a bug, which has a higher priority than a feature. The bug is assigned to the relevant internal group. If our triage gets approved will be scheduled for resolution. I don't have an ETA yet. I'll keep this issue updated with new information as soon as I have them.

**Comment by fadilafidina** [comment]
Thanks for the update @mauroservienti.

In the meantime, we will turn compatibility mode back on and set a MessageVisibilityTimeout value there. We would love to get off it once a new release is out. Would there be any other workarounds that you'd recommend?

As a side question, the processing that's taking a while is happening in a saga handler, would it be ideal to move it into a regular handler, as mentioned here? https://docs.particular.net/nservicebus/sagas/concurrency#high-load-scenarios-avoid-io-or-cpu-bound-operations-in-a-saga-handler

**Comment by mauroservienti** [comment]
> In the meantime, we will turn compatibility mode back on and set a MessageVisibilityTimeout value there. We would love to get off it once a new release is out. Would there be any other workarounds that you'd recommend?

Another option, that is probably simple as it doesn't require code changes is to go back to version 5.3. Looking at the [diff](https://github.com/Particular/NServiceBus.AmazonSQS/compare/release-5.3...release-5.4) the only difference is the implementation of in-memory rate limiting and caching to support high throughput scenarios that require the compatibility scenario.

If you decide to use this second workaround, I recommend some testing before going to production.

**Comment by mauroservienti** [comment]
> As a side question, the processing that's taking a while is happening in a saga handler, would it be ideal to move it into a regular handler, as mentioned here? https://docs.particular.net/nservicebus/sagas/concurrency#high-load-scenarios-avoid-io-or-cpu-bound-operations-in-a-saga-handler

Moving long-running processes outside of saga handlers is beneficial for sagas with high concurrency. It would be prevalent for the long-running handler to fail and be retried because someone else touched the saga in the meantime.

**Comment by fadilafidina** [comment]
I think moving down a version for us is more work than adding compatibility mode, so we will go with that. Thanks for the alternative solution.

We look forward to this being fixed! Thanks.

**Comment by mauroservienti** [comment]
@fadilafidina, an update on the timeline. The issue has been scheduled for resolution. It will be assigned to a group soon. I don't have an ETA, but we're talking about days, not weeks.

**Comment by fadilafidina** [comment]
Thanks @mauroservienti for the update.

**Comment by mauroservienti** [comment]
@fadilafidina thanks for your patience. @soujay and I started working on this today.
We raised three pull requests to fix the bug on [5.6.1](https://github.com/Particular/NServiceBus.AmazonSQS/pull/1235) and backport it to [5.5.1](https://github.com/Particular/NServiceBus.AmazonSQS/pull/1236) and [5.4.1](https://github.com/Particular/NServiceBus.AmazonSQS/pull/1237)

**Comment by soujay** [comment]
@fadilafidina - We just released patches to the latest version that addresses the bug. You can install the new version of NServiceBus.AmazonSQS from [NuGet](https://www.nuget.org/packages/NServiceBus.AmazonSQS).
